src/System.Private.CoreLib/src/System/Threading/ManagedThreadId.cs

```diff
@@ -9,15 +9,171 @@
 // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
 
 using System;
-using System.Diagnostics.Contracts;
+using System.Diagnostics;
 using System.Runtime;
-using System.Runtime.InteropServices;
-using System.Security;
 
 namespace System.Threading
 {
     internal class ManagedThreadId
     {
+        //
+        // Binary tree used to keep track of active thread ids. Each node of the tree keeps track of 32 consecutive ids.
+        // Implemented as immutable collection to avoid locks. Each modification creates a new top level node.
+        // 
+        class ImmutableIdDispenser
+        {
+            private readonly ImmutableIdDispenser _left; // Child nodes
+            private readonly ImmutableIdDispenser _right;
+
+            private readonly int _used; // Number of ids tracked by this node and all its childs
+            private readonly int _size; // Maximum number of ids that can be tracked by this node and all its childs
+
+            private readonly uint _bitmap; // Bitmap of ids tracked by this node
+
+            private const int BitsPerNode = 32;
```

Comment:
I am ok with the current implementation as long it satisfy the perf requirements as you mentioned,  I'll talk to you in person to know if it is worth trying other mechanisms but this should stop you from merging your pull request